aezeed/cipherseed.go

```diff
@@ -0,0 +1,542 @@
+	// CipherSeed struct.
+	var c CipherSeed
+	if err := c.decode(bytes.NewReader(plainSeed[:])); err != nil {
+		return nil, err
+	}
+
+	return &c, nil
+}
+
+// decipherCipherSeed attempts to decipher the passed cipher seed ciphertext
+// using the passed passphrase. This function is the opposite of
+// the encipher method.
+func decipherCipherSeed(cipherSeedBytes [EncipheredCipherSeedSize]byte,
+	pass []byte) ([DecipheredCipherSeedSize]byte, error) {
+
+	var plainSeed [DecipheredCipherSeedSize]byte
+
+	// Before we do anything, we'll ensure that the version is one that we
+	// understand. Otherwise, we won't be able to decrypt, or even parse
+	// the cipher seed.
+	if uint8(cipherSeedBytes[0]) != CipherSeedVersion {
+		return plainSeed, ErrIncorrectVersion
+	}
+
+	// Next, we'll slice off the salt from the pass cipher seed, then
+	// snip off the end of the cipher seed, ignoring the version, and
+	// finally the checksum.
+	salt := cipherSeedBytes[saltOffset : saltOffset+saltSize]
+	cipherSeed := cipherSeedBytes[1:saltOffset]
+	checksum := cipherSeedBytes[checkSumOffset:]
+
+	// Before we perform any crypto operations, we'll remote and verify the
+	// checksum to ensure that the user input the proper set of words.
+	freshChecksum := crc32.Checksum(cipherSeedBytes[:checkSumOffset], crcTable)
+	if freshChecksum != binary.BigEndian.Uint32(checksum) {
+		return plainSeed, ErrIncorrectMnemonic
+	}
+
+	// With the salt separated from the cipher text, we'll now obtain the
+	// key used for encryption.
+	key, err := scrypt.Key(pass, salt, scryptN, scryptR, scryptP, keyLen)
+	if err != nil {
+		return plainSeed, err
+	}
+
+	// We'll also extract the AD that will be required to properly pass the
+	// MAC check.
+	ad := extractAD(cipherSeedBytes)
+
+	// With the key, we'll attempt to decrypt the plaintext. If the
+	// ciphertext was altered, or the passphrase is incorrect, then we'll
+	// error out.
+	plainSeedBytes, ok := aez.Decrypt(
+		key, nil, [][]byte{ad[:]}, CipherTextExpansion, cipherSeed, nil,
+	)
+	if !ok {
+		return plainSeed, ErrInvalidPass
+	}
+	copy(plainSeed[:], plainSeedBytes)
+
+	return plainSeed, nil
+
+}
+
+// Decipher attempts to decipher the encoded mnemonic by first mapping to the
+// original chipertext, then applying our deciphering scheme. ErrInvalidPass
+// will be returned if the passphrase is incorrect.
+func (m *Mnemonic) Decipher(pass []byte) ([DecipheredCipherSeedSize]byte, error) {
+
+	// Before we attempt to map the mnemonic back to the original
+	// ciphertext, we'll ensure that all the word are actually a part of
+	// the current default word list.
+	for _, word := range m {
+		if !strings.Contains(englishWordList, word) {
+			emptySeed := [DecipheredCipherSeedSize]byte{}
+			return emptySeed, ErrUnknownMnenomicWord{word}
+		}
+	}
+
+	// If the passphrase wasn't provided, then we'll use the string
+	// "aezeed" in place.
+	passphrase := pass
+	if len(passphrase) == 0 {
+		passphrase = []byte("aezeed")
```

Comment:
extract the default passphrase into constant?

Reply:
Fixed.